experiments/conductor/cmd/runner/utilities.go

```diff
@@ -296,61 +298,90 @@ func executeCommand(opts *RunnerOptions, cfg CommandConfig) (string, string, err
 	}
-	if len(cfg.Env) > 0 {
-		log.Printf("[%s] environment:", cfg.Name)
-		for k, v := range cfg.Env {
-			log.Printf("  %s=%s", k, v)
-		}
+
+	maxRetries := opts.defaultRetries
+	// If MaxRetries is set in config, cap it at that
+	if cfg.MaxRetries > 0 && cfg.MaxRetries < maxRetries {
+		maxRetries = cfg.MaxRetries
 	}
 
-	ctx, cancel := context.WithTimeout(context.Background(), cfg.Timeout)
-	defer cancel()
+	var lastErr error
+	var output, errOutput string
+	retryCount := 0
 
-	cmd := exec.CommandContext(ctx, cfg.Cmd, cfg.Args...)
-	cmd.Dir = cfg.WorkDir
+	for {
+		log.Printf("Starting command step: %s (attempt %d/%d)", cfg.Name, retryCount+1, maxRetries+1)
+		log.Printf("[%s] working directory: %s", cfg.Name, cfg.WorkDir)
+		log.Printf("[%s] command: %s %s", cfg.Name, cfg.Cmd, strings.Join(cfg.Args, " "))
+		if cfg.Stdin != nil {
+			log.Printf("[%s] stdin: %s", cfg.Name, cfg.Stdin)
+		}
+		if len(cfg.Env) > 0 {
+			log.Printf("[%s] environment:", cfg.Name)
+			for k, v := range cfg.Env {
+				log.Printf("  %s=%s", k, v)
+			}
+		}
 
-	var outBuf, errBuf strings.Builder
-	cmd.Stdout = &outBuf
-	cmd.Stderr = &errBuf
+		ctx, cancel := context.WithTimeout(context.Background(), cfg.Timeout)
+		cmd := exec.CommandContext(ctx, cfg.Cmd, cfg.Args...)
+		cmd.Dir = cfg.WorkDir
 
-	if cfg.Stdin != nil {
-		cmd.Stdin = cfg.Stdin
-	}
+		var outBuf, errBuf strings.Builder
+		cmd.Stdout = &outBuf
+		cmd.Stderr = &errBuf
 
-	// Set up environment variables
-	if len(cfg.Env) > 0 {
-		cmd.Env = os.Environ() // Start with current environment
-		for k, v := range cfg.Env {
-			cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
+		if cfg.Stdin != nil {
+			cmd.Stdin = cfg.Stdin
+		}
+
+		// Set up environment variables
+		if len(cfg.Env) > 0 {
+			cmd.Env = os.Environ() // Start with current environment
+			for k, v := range cfg.Env {
+				cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
+			}
 		}
-	}
 
-	start := time.Now()
+		start := time.Now()
+		err := cmd.Run()
+		stop := time.Now()
+		diff := stop.Sub(start)
+		cancel()
+
+		output = outBuf.String()
+		errOutput = errBuf.String()
+
+		if err == nil {
+			log.Printf("[%s] SUCCESS (%v)", cfg.Name, diff)
+			printCommandOutput(output)
+			if errBuf.Len() > 0 {
+				log.Printf("[%s] stderr output:", cfg.Name)
+				printCommandOutput(errOutput)
+			}
+			return output, errOutput, nil
+		}
 
-	err := cmd.Run()
-	stop := time.Now()
-	diff := stop.Sub(start)
-	if err != nil {
-		log.Printf("[%s] ERROR (%v): \n", cfg.Name, diff)
-		log.Printf("[%s] err: %q\n", cfg.Name, err)
-	} else {
-		log.Printf("[%s] SUCCESS (%v): \n", cfg.Name, diff)
-	}
+		lastErr = err
+		log.Printf("[%s] ERROR (%v): %v", cfg.Name, diff, err)
+		printCommandOutput(output)
+		if errBuf.Len() > 0 {
+			log.Printf("[%s] stderr output:", cfg.Name)
+			printCommandOutput(errOutput)
+		}
 
-	output := outBuf.String()
-	errOutput := errBuf.String()
+		retryCount++
+		if maxRetries >= 0 && retryCount > maxRetries {
```

Comment:
Nit: Looks like the maxRetries is either opts.defaultRetries or the hard coded maxRetries, so `if retryCount > maxRetries {` seems to be sufficient. Otherwise when opts.defaultRetries is set to a negative value, then this command will never stop by itself.

Reply:
added a check to make sure defaultretries is not -ve